File: 프로그래머스/2/42587. 프로세스/프로세스.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        Queue<(int, int)> q = new Queue<(int, int)>();
        
        int idx = 0;
        foreach(var i in priorities)
        {
            q.Enqueue((idx++, i));
        }
        
        while(true)
        {
            var max = q.Max(x => x.Item2);
            var cur = q.Dequeue();
            if(cur.Item2 == max)
            {
                answer++;
                if(cur.Item1 == location)
                {
                    return answer;
                }
            }
            else
            {
                q.Enqueue(cur);
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    queue<pair<int, int>> q;
    priority_queue<int> pq;
    
    for(int i = 0; i < priorities.size(); i++)
    {
        q.push({priorities[i], i});
        pq.push(priorities[i]);
    }
    
    while(q.size())
    {
        int curPriority = q.front().first;
        int curIdx = q.front().second;
        q.pop();
        
        if(curPriority == pq.top())
        {
            answer++;
            pq.pop();
            if(curIdx == location)
                return answer;
        }
        else
        {
            q.push({curPriority, curIdx});
        }
    }
    return answer;
}